### PR TITLE
[DebuggerV2] Add types, selectors, reducers & effects for source code

### PR DESCRIPTION
**Pull request by caisq**
* Motivation for features / changes
  * Continue developing DebuggerV2 plugin.

* Technical description of changes
  * Add some supporting types for source code and stack frames
  * Add data source method for fetching list of all source files known to the debugger
  * Add selectors for the source-file-list-related state fields
  * Add actions for requesting and loading the source file list
  * Add effects for loading the source-file list on debugger view load

* Screenshots of UI changes
  * N/A since this is a pure type + ngrx change

* Detailed steps to verify changes work correctly (as executed by you)
  * Unit tests added

### CODE REVIEW

**Comment by caisq** [comment]
cc @stephanwlee @baileydesign

**Comment by caisq** [comment]
Gentle ping @psybuzz :) Please again note that this CL is orthogonal to the ongoing design discussion of the third-party library for actual code rendering.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts [hunk 1]
```diff
@@ -469,6 +478,37 @@ const reducer = createReducer(
...
+  ),
+  on(
+    actions.sourceFileListLoaded,
+    (state: DebuggerState, sourceFileList): DebuggerState => {
+      return {
+        ...state,
+        sourceCode: {
+          ...state.sourceCode,
+          sourceFileListLoaded: {
+            state: DataLoadState.LOADED,
```
Comment (psybuzz) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts]: nit: since L489 adds 
```
sourceFileListLoaded: {
            ...state.sourceCode.sourceFileListLoaded,
```
we should probably add the same thing here for safety?

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_types.ts [hunk 2]
```diff
@@ -206,6 +206,26 @@ export interface RunState {
...
 }
 
+// A source-code file.
+export interface SourceFileSpec {
+  host_name: string;
+  file_path: string;
+}
+
+// A specific line of a source-code file, i.e., a stack frame.
+export interface SourceLineSpec extends SourceFileSpec {
```
Comment (psybuzz) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_types.ts]: nit: totally optional and up to you, but for naming, wdyt about `SourceLocation` ?

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts [hunk 3]
```diff
@@ -469,6 +478,37 @@ const reducer = createReducer(
...
+            lastLoadedTimeInMs: Date.now(),
+          },
+          sourceFileList: sourceFileList.sourceFiles,
```
Comment (psybuzz) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts]: Based on the data access, we always pass a `run` when requesting sourceFiles.

To double check: is there any use case in the UX when a user wants to keep seeing stale source files?  or source files from a previously selected run?  (e.g. should we just merge the new + previous list of source files?)

I'd imagine if there's a run-selector UI control, a user might switch back and forth between 2 runs many times, and since we don't index the sourceFiles by run-id, we would need to re-fetch on each switch.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/data_source/tfdbg2_data_source.ts [hunk 4]
```diff
@@ -25,6 +25,8 @@ import {TBHttpClient} from '../../../../webapp/webapp_data_source/tb_http_client
 
 /** @typehack */ import * as _typeHackRxjs from 'rxjs';
 
+export type SourceFileListResponse = Array<[string, string]>;
```
[the remark this answers is not in the document]
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/data_source/tfdbg2_data_source.ts]: Added comment to clarify what the two strings are here.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts [hunk 5]
```diff
@@ -608,13 +613,44 @@ export class DebuggerEffects {
     );
   }
 
+  /**
+   * Load list of source files when debugger plugin is loaded.
+   */
+  private loadSourceFileList(prevStream$: Observable<void>) {
```
[the remark this answers is not in the document]
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: Acknowledged. Your understanding is correct.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts [hunk 6]
```diff
@@ -608,13 +613,44 @@ export class DebuggerEffects {
...
+    return prevStream$.pipe(
+      withLatestFrom(
+        this.store.select(getActiveRunId),
+        this.store.select(getSourceFileListLoaded)
+      ),
+      filter(([, runId, fileIndexLoadState]) => {
```
[the remark this answers is not in the document]
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: Agreed. I made revisions to make the naming more consistent.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts (hunk 1 above)
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts]: Done.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_types.ts (hunk 2 above)
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_types.ts]: I prefer keeping `SourceLineSpec`. Reasons:

1. It's symmetric w.r.t. `SourceFileSpec` above.
2. It's specific about whether it's talking about a line or a range of lines (the former is the case currently), and whether it's talking about a single stack frame or a list of them (again the former is the case.)

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts (hunk 3 above)
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts]: Good question. The debugger v2 FE currently is implemented in a way that shows only one run at a time. This keeps the data structure simpler than what it would be if multiple runs are handled. This is sufficient IMO because when during TF debugging, you typically examine one run (i.e., one training job) at a time. There is less concern for comparing, say the loss curves from different hyperparameters from different runs, than in other TB dashboards such as scalars.
